### ISSUE TITLE
Zarf - Referrer can frontrun purchases and maximise their fee

### ISSUE THREAD

**Issue by sherlock-admin** [sherlock-audit/2022-11-bond-judging#24]
Zarf

medium

# Referrer can frontrun purchases and maximise their fee

## Summary

There is no way for the user to be sure the visible referrer fee is the actual fee which has to be paid upon purchasing bonds. A referrer can frontrun bond purchases and maximise their fee (5% of the amount sent in). Only after the purchase has been successfully performed, the user knows  how much fee they paid to the referrer.

## Vulnerability Detail

Imagine the current referrer fee for a specific referrer is 0.1% of the amount used to purchase bonds. The referrer might monitor the mempool for those specific contract calls which include their address as the referrer in `purchase()` of the `BondBaseTeller` contract. 

If one of those transactions are residing in the mempool, the referrer could create a transaction with a higher gas price/fee to set the referrer fee to 5% using the `setReferrerFee()` in the `BondBaseTeller` contract. This ensures the fee is set to 5% before the purchase of the user will be performed.

Next, the transaction of the user will be executed, which will purchase bonds for a specific amount of quote tokens. However, first the referrer fee (max 5%) will be deducted from amount, resulting in a loss for the user and benefit for the referrer. 

Only after the transaction is confirmed and the user received his/her bonds, it’s clear the referrer took 5% of the sent quote tokens (instead of the 0.1% as shown prior to the purchase).

## Impact

Users might receive less tokens as expected when purchasing bonds 

## Code Snippet

[https://github.com/sherlock-audit/2022-11-bond/blob/main/src/bases/BondBaseTeller.sol#L88-L91](https://github.com/sherlock-audit/2022-11-bond/blob/main/src/bases/BondBaseTeller.sol#L88-L91)

[https://github.com/sherlock-audit/2022-11-bond/blob/main/src/bases/BondBaseTeller.sol#L121-L166](https://github.com/sherlock-audit/2022-11-bond/blob/main/src/bases/BondBaseTeller.sol#L121-L166)

## Tool used

Manual Review

## Recommendation

Either make the referrer fee immutable, such that the user can be sure the current fee does not increase prior to purchasing bonds. 

Alternatively a timelock could be introduced to change the fee amount. This way, frontrunning wouldn't be possible and users would know be certain on the fees they are agreeing with.

Duplicate of #29
